File: Ejercicio 19/Ejercicio 19.py
# Clase Nodo para la lista enlazada
class Nodo:
    def __init__(self, valor):
        self.dato = valor
        self.siguiente = None

# Clase Pila implementada con lista enlazada
class Pila:
    """
    @brief Constructor de la pila
    @param cap Capacidad máxima de la pila
    """
    def __init__(self, cap):
        self.top = None       # Referencia al tope de la pila
        self.capacidad = cap  # Capacidad máxima de la pila
        self.tamaño = 0       # Tamaño actual de la pila
        
        print("\n")
        print(f"Pila creada con capacidad: {self.capacidad}")

    """
    @brief Verifica si la pila está completamente llena
    @return true si la pila está llena, false en caso contrario
    """
    def is_full(self):
        return self.tamaño == self.capacidad

    """
    @brief Verifica si la pila está vacía
    @return true si la pila está vacía, false en caso contrario
    """
    def is_empty(self):
        return self.top is None

    """
    @brief Inserta un elemento en la cima de la pila (operación PUSH)
    @param valor El valor a insertar
    """
    def push(self, valor):
        if self.is_full():
            print(f"ERROR: Desbordamiento de pila (Stack Overflow). No se puede insertar '{valor}'.")
        else:
            # Crear nuevo nodo
            nuevo_nodo = Nodo(valor)
            
            # Insertar al inicio (tope de la pila)
            nuevo_nodo.siguiente = self.top
            self.top = nuevo_nodo
            self.tamaño += 1
            
            print(f"PUSH: Elemento '{valor}' insertado en la pila.")

    """
    @brief Elimina y devuelve el elemento de la cima de la pila (operación POP)
    @return El elemento eliminado, o cadena vacía si la pila está vacía
    """
    def pop(self):
        if self.is_empty():
            print("ERROR: Desbordamiento negativo de pila (Stack Underflow). No hay elementos para eliminar.")
            return ""  # Cadena vacía para indicar error
        else:
            # Recuperar el elemento del tope
            valor_eliminado = self.top.dato
            
            # Eliminar el nodo del tope
            temp = self.top
            self.top = self.top.siguiente
            # En Python el garbage collector se encarga de liberar la memoria
            self.tamaño -= 1
            
            print(f"POP: Elemento '{valor_eliminado}' eliminado de la pila.")
            return valor_eliminado

    """
    @brief Devuelve el elemento de la cima sin eliminarlo (operación PEEK)
    @return El elemento de la cima, o cadena vacía si la pila está vacía
    """
    def peek(self):
        if self.is_empty():
            print("ERROR: La pila está vacía. No hay elementos para ver (Peek).")
            return ""  # Cadena vacía para indicar error
        else:
            return self.top.dato

    """
    @brief Muestra todos los elementos de la pila
    """
    def mostrar(self):
        if self.is_empty():
            print("La pila está vacía.")
        else:
            print("Elementos en la pila (desde el tope):")
            actual = self.top
            posicion = 1
            
            while actual is not None:
                print(f"{posicion}. {actual.dato}")
                actual = actual.siguiente
                posicion += 1

    """
    @brief Obtiene el tamaño actual de la pila
    @return Número de elementos en la pila
    """
    def get_tamaño(self):
        return self.tamaño

    """
    @brief Obtiene la capacidad máxima de la pila
    @return Capacidad máxima
    """
    def get_capacidad(self):
        return self.capacidad

"""
@brief Función para mostrar el menú de opciones
"""
def mostrar_menu():
    print("\n---- MENU DE OPERACIONES ---")
    print("1. PUSH (Insertar elemento)")
    print("2. POP (Eliminar elemento)")
    print("3. PEEK (Ver elemento superior)")
    print("4. MOSTRAR (Mostrar todos los elementos)")
    print("5. ESTADO (Ver estado de la pila)")
    print("6. SALIR")
    print("-----------------------------")
    print("\n")

"""
@brief Función principal
"""
def main():
    print("--- Pila con Lista Enlazada Simple ---")
    capacidad_input = input("Ingrese la capacidad de la pila: ")
    capacidad = int(capacidad_input)
    
    # Crear la pila
    mi_pila = Pila(capacidad)
    
    while True:
        mostrar_menu()
        opcion_input = input("Seleccione una opción: ")
        opcion = int(opcion_input)
        
        if opcion == 1:  # PUSH
            print("\n")
            valor = input("Ingrese el valor a insertar: ")
            mi_pila.push(valor)
            print("\n")
            
        elif opcion == 2:  # POP
            print("\n")
            mi_pila.pop()
            
        elif opcion == 3:  # PEEK
            print("\n")
            tope = mi_pila.peek()
            if tope is not None and tope != "":
                print(f"Elemento en el tope: {tope}")
            print("\n")
            
        elif opcion == 4:  # MOSTRAR
            print("\n")
            mi_pila.mostrar()
            
        elif opcion == 5:  # ESTADO
            print("\n--- ESTADO DE LA PILA ---")
            print(f"Capacidad máxima: {mi_pila.get_capacidad()}")
            print(f"Elementos actuales: {mi_pila.get_tamaño()}")
            print(f"¿Está vacía? {'Sí' if mi_pila.is_empty() else 'No'}")
            print(f"¿Está llena? {'Sí' if mi_pila.is_full() else 'No'}")
            print("\n")
            
        elif opcion == 6:  # SALIR
            print("\n")
            print("Saliendo del programa...")
            break
            
        else:
            print("Opción no válida. Intente nuevamente.")

# Ejecutar el programa
if __name__ == "__main__":
    main()

File: Ejercicio 19/Ejercicio 19.js
// Clase Nodo para la lista enlazada
class Nodo {
    constructor(valor) {
        this.dato = valor;
        this.siguiente = null;
    }
}

// Clase Pila implementada con lista enlazada
class Pila {
    /**
     * @brief Constructor de la pila
     * @param cap Capacidad máxima de la pila
     */
    constructor(cap) {
        this.top = null;      // Referencia al tope de la pila
        this.capacidad = cap; // Capacidad máxima de la pila
        this.tamaño = 0;      // Tamaño actual de la pila
        
        console.log("\n");
        console.log(`Pila creada con capacidad: ${this.capacidad}`);
    }

    /**
     * @brief Verifica si la pila está completamente llena
     * @return true si la pila está llena, false en caso contrario
     */
    isFull() {
        return this.tamaño === this.capacidad;
    }

    /**
     * @brief Verifica si la pila está vacía
     * @return true si la pila está vacía, false en caso contrario
     */
    isEmpty() {
        return this.top === null;
    }

    /**
     * @brief Inserta un elemento en la cima de la pila (operación PUSH)
     * @param valor El valor a insertar
     */
    push(valor) {
        if (this.isFull()) {
            console.log(`ERROR: Desbordamiento de pila (Stack Overflow). No se puede insertar '${valor}'.`);
        } else {
            // Crear nuevo nodo
            const nuevoNodo = new Nodo(valor);
            
            // Insertar al inicio (tope de la pila)
            nuevoNodo.siguiente = this.top;
            this.top = nuevoNodo;
            this.tamaño++;
            
            console.log(`PUSH: Elemento '${valor}' insertado en la pila.`);
        }
    }

    /**
     * @brief Elimina y devuelve el elemento de la cima de la pila (operación POP)
     * @return El elemento eliminado, o cadena vacía si la pila está vacía
     */
    pop() {
        if (this.isEmpty()) {
            console.log("ERROR: Desbordamiento negativo de pila (Stack Underflow). No hay elementos para eliminar.");
            return ""; // Cadena vacía para indicar error
        } else {
            // Recuperar el elemento del tope
            const valorEliminado = this.top.dato;
            
            // Eliminar el nodo del tope
            const temp = this.top;
            this.top = this.top.siguiente;
            // En JavaScript el garbage collector se encarga de liberar la memoria
            this.tamaño--;
            
            console.log(`POP: Elemento '${valorEliminado}' eliminado de la pila.`);
            return valorEliminado;
        }
    }

    /**
     * @brief Devuelve el elemento de la cima sin eliminarlo (operación PEEK)
     * @return El elemento de la cima, o cadena vacía si la pila está vacía
     */
    peek() {
        if (this.isEmpty()) {
            console.log("ERROR: La pila está vacía. No hay elementos para ver (Peek).");
            return ""; // Cadena vacía para indicar error
        } else {
            return this.top.dato;
        }
    }

    /**
     * @brief Muestra todos los elementos de la pila
     */
    mostrar() {
        if (this.isEmpty()) {
            console.log("La pila está vacía.");
        } else {
            console.log("Elementos en la pila (desde el tope):");
            let actual = this.top;
            let posicion = 1;
            
            while (actual !== null) {
                console.log(`${posicion}. ${actual.dato}`);
                actual = actual.siguiente;
                posicion++;
            }
        }
    }

    /**
     * @brief Obtiene el tamaño actual de la pila
     * @return Número de elementos en la pila
     */
    getTamaño() {
        return this.tamaño;
    }

    /**
     * @brief Obtiene la capacidad máxima de la pila
     * @return Capacidad máxima
     */
    getCapacidad() {
        return this.capacidad;
    }
}

// Para simular la entrada del usuario en el navegador
const readline = require('readline');

/**
 * @brief Función para mostrar el menú de opciones
 */
function mostrarMenu() {
    console.log("\n---- MENU DE OPERACIONES ---");
    console.log("1. PUSH (Insertar elemento)");
    console.log("2. POP (Eliminar elemento)");
    console.log("3. PEEK (Ver elemento superior)");
    console.log("4. MOSTRAR (Mostrar todos los elementos)");
    console.log("5. ESTADO (Ver estado de la pila)");
    console.log("6. SALIR");
    console.log("-----------------------------");
    console.log("\n");
}

/**
 * @brief Función principal
 */
async function main() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    // Función auxiliar para preguntar al usuario
    const pregunta = (query) => new Promise((resolve) => rl.question(query, resolve));

    try {
        console.log("--- Pila con Lista Enlazada Simple ---");
        const capacidadInput = await pregunta("Ingrese la capacidad de la pila: ");
        const capacidad = parseInt(capacidadInput);
        
        // Crear la pila
        const miPila = new Pila(capacidad);
        
        let opcion;
        
        do {
            mostrarMenu();
            const opcionInput = await pregunta("Seleccione una opción: ");
            opcion = parseInt(opcionInput);
            
            switch (opcion) {
                case 1: // PUSH
                    console.log("\n");
                    const valor = await pregunta("Ingrese el valor a insertar: ");
                    miPila.push(valor);
                    console.log("\n");
                    break;
                    
                case 2: // POP
                    console.log("\n");
                    miPila.pop();
                    break;
                    
                case 3: // PEEK
                    console.log("\n");
                    const tope = miPila.peek();
                    if (tope !== null && tope !== "") {
                        console.log(`Elemento en el tope: ${tope}`);
                    }
                    console.log("\n");
                    break;
                    
                case 4: // MOSTRAR
                    console.log("\n");
                    miPila.mostrar();
                    break;
                    
                case 5: // ESTADO
                    console.log("\n--- ESTADO DE LA PILA ---");
                    console.log(`Capacidad máxima: ${miPila.getCapacidad()}`);
                    console.log(`Elementos actuales: ${miPila.getTamaño()}`);
                    console.log(`¿Está vacía? ${miPila.isEmpty() ? "Sí" : "No"}`);
                    console.log(`¿Está llena? ${miPila.isFull() ? "Sí" : "No"}`);
                    console.log("\n");
                    break;
                    
                case 6: // SALIR
                    console.log("\n");
                    console.log("Saliendo del programa...");
                    break;
                    
                default:
                    console.log("Opción no válida. Intente nuevamente.");
                    break;
            }
            
        } while (opcion !== 6);
        
    } finally {
        rl.close();
    }
}

// Ejecutar el programa
if (typeof window === 'undefined') {
    // Entorno Node.js
    main().catch(console.error);
} 